core/modules/wpublish/qserv-worker-perf-chunks.cc

```diff
@@ -46,101 +160,79 @@ int test() {
-                    value,
-                    [&finished] (wpublish::TestEchoQservRequest::Status status,
-                                 string const& error,
-                                 string const& sent,
-                                 string const& received) {
-
-                        if (status != wpublish::TestEchoQservRequest::Status::SUCCESS) {
-                            cout << "status: " << wpublish::TestEchoQservRequest::status2str(status) << "\n"
-                                 << "error:  " << error << endl;
-                        } else {
-                            cout << "value sent:     " << sent     << "\n"
-                                 << "value received: " << received << endl;
-                        }
-                        finished--;
-                    });
-                requests.push_back(request);
-
-                // Submit the request
-                finished++;
-                XrdSsiResource resource(resourcePath);
-                serviceProvider->ProcessRequest(*request, resource);
+                auto&& resourcePath = resources[j];
+                jobs.push_back(resourcePath);
             }
         }
     }
-    if (cancelAfterMs == 0) {
-        // Block while at least one request is in progress
-        util::BlockPost blockPost(1000, 2000);
-        while (finished) {
-            blockPost.wait(200);
-        }
-    } else {
-        // Request cancellation timeout is used to test the correctness of
-        // the XRootD/SSI implementation under heavy loads.
-        util::BlockPost blockPost(cancelAfterMs, cancelAfterMs+1);
-        blockPost.wait();
-        for (auto&& request: requests) {
-            bool const cancel = true;
-            request->Finished(cancel);
-        }
+
+    // Make allocation of jobs to threads using the 'round-robin' method
+    vector<vector<string>> thread2jobs(numThreads);
+    for (unsigned int i = 0; i < jobs.size(); ++i) {
+        auto const threadIdx = i % numThreads;
+        auto&& resourcePath = jobs[i];
+        thread2jobs[threadIdx].push_back(resourcePath);
+    }
+
+    Counter numRequestsInFlight(flowControlLimit);
+
+    // Launch threads
+    vector<unique_ptr<thread>> threads;
+    for (unsigned int tIdx = 0; tIdx < numThreads; ++tIdx) {
+
+        threads.emplace_back(new thread(
+            [&numRequestsInFlight](vector<string> const& jobs,
+                                   XrdSsiService* serviceProvider) {
+                // Launch requests
+                vector<RequestT::Ptr> requests;
+                for (auto&& resourcePath: jobs) {
+                    requests.push_back(makeRequest(numRequestsInFlight, serviceProvider, resourcePath));
+                }
+    
+                // Block while at least one request is in progress
+                util::BlockPost blockPost(1000, 2000);
+                while (numRequestsInFlight.counter() != 0) {
+                    blockPost.wait(200);
+                }
```

Comment:
Why the random wait time? It looks like it is trying to hide a potential race condition with numRequestedInFlight.
I think you would be better off with a separate Counter defined in this thread, say localCounter. localCounter would be incremented in the requests.push_back loop (the increment must happen in this thread or there's a race condition) and then decremented at the same point numRequestsInFlight is decremented (so it needs to be passed to makeRequest). Then just wait for localCounter to reach 0.

Reply:
Good point! Thank you!

Reply:
On the other hand, I think the join below is enough since there isn't any other logic that needs to be run before the join.